dali/kernels/erase/erase_gpu.h

```diff
@@ -0,0 +1,530 @@
+ *
+ * We cover every n-dimensional sample by n-dimensional cubes (in fact a hyperrectangle)
+ * of one selected size. We call such a hyperrectangle a region.
+ *
+ * The size of such region is suitable for traversal of block of threads of dimension (32, 32, 1)
+ *
+ *
+ * Grid: (number of cubes, number of samples, 1)
+ * - the number of cubes is the max over the numbers of cubes to cover each sample
+ *
+ * Block: (32, 32, 1) - the dimensions of thread block are fixed. We use such "physical block"
+ * to traverse every region - the hyperrectangle.
+ *
+ * Steps:
+ * 1. In every block we first look for intersections with erased regions, to filter out only
+ * the ones that we actually need.
+ * 2. We check if the hyperrectangle is fully covered by erase region or not covered at all
+ * to delegate to special cases - fill only or memcopy
+ * 3. If we have mixed cover, we check each coordinate and either copy the input or the fill value
+ *
+ * The processing of hyperrectangle of n-dims is done by n-nested for loop from outer to innermost
+ * dimension.
+ * The two innermost dimensions (omitting the channel) are traversed by the block of parallel
+ * threads.
+ * The loop is implemented as template based recursion.
+ * Possible optimization/obfuscation - factor out checks over dimension to specific level of
+ * the loop nest (for the mixed cover case).
+ *
+ */
+
+template <typename T, int ndim>
+using box = Box<ndim, T>;
+
+template <int ndim>
+using ibox = box<int32_t, ndim>;
+
+template <typename T, int ndim>
+struct erase_sample_desc {
+  const T *in = nullptr;
+  T* out = nullptr;
+  const ibox<ndim> *erase_regions = nullptr;
+  ivec<ndim> sample_shape;
+  ivec<ndim> sample_stride;
+};
+
+template <int ndim>
+DALI_HOST_DEV ivec<ndim> get_region_start(int region_idx, ivec<ndim> region_shape,
+                                          ivec<ndim> sample_shape) {
+  // let's calculate how many regions cover given axes:
+  auto region_cover_shape = div_ceil(sample_shape, region_shape);
+  // this introduces the same ordering of regions as data layout
+  auto region_cover_strides = GetStrides(region_cover_shape);
+  ivec<ndim> region_position{};  // n-dim index of region among other regions
+  region_position[0] = region_idx / region_cover_strides[0];
+  for (int i = 1; i < ndim; i++) {
+    region_idx -= region_position[i - 1] * region_cover_strides[i - 1];
+    region_position[i] = region_idx / region_cover_strides[i];
+  }
+  return region_position * region_shape;
+}
+
+template <int ndim>
+__device__ ibox<ndim> get_region(int region_idx, ivec<ndim> region_shape, ivec<ndim> sample_shape) {
+  auto anchor = get_region_start(region_idx, region_shape, sample_shape);
+  return {anchor, anchor + region_shape};
+}
+
+constexpr unsigned FULL_MASK = 0xffffffffu;
+
+/**
+ * @brief Returns true if that parameter configuration is just outer layer of recursive loop call
+ */
+constexpr bool select_outer_loops(int ndim, int current_dim, int channel_dim) {
```

Comment:
Nitpick (?): the name doesn't explain much... I think that something with recurse in the name would be more appropriate - after all, don't you always loop over the outer dimension?

Reply:
How about
```suggestion
constexpr bool is_outer_loop(int ndim, int current_dim, int channel_dim) {
```